rig/routing_table/minimise.py

```diff
@@ -0,0 +1,557 @@
+            ones |= entry.key
+            key &= entry.key
+            mask &= entry.mask
+
+        # Compute the new mask, key and generality
+        # NOTE: `~key` indicates wherever there is a `0` in any key.
+        # Consequently `ones ^ ~key` indicates any bits in which there are
+        # either only zeros or ones (or, the bits where there doesn't need to
+        # be an X in the new entry).
+        mask &= (ones ^ ~key)  # Add any new Xs
+        key &= mask
+        generality = get_generality(key, mask)
+        insertion_index = get_insertion_index(routing_table, generality)
+
+        # Compute the goodness of the merge
+        goodness = len(entries) - 1
+
+        return super(Merge, cls).__new__(
+            cls, routing_table, frozenset(entries), key, mask,
+            generality, goodness, insertion_index
+        )
+
+    def apply(self, aliases):
+        """Apply the merge to the routing table it is defined against and get a
+        new routing table and alias dictionary.
+
+        Returns
+        -------
+        [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+            A new routing table which may be minimised further.
+        {(key, mask): {(key, mask), ...}}
+            A new dictionary mapping keys and masks from the new routing table
+            to sets of keys and masks that they represent.
+        """
+        # Generate a new empty table of the correct size
+        routing_entries = list(self.routing_table)
+        aliases = dict(aliases)
+
+        # Add the new entry
+        new_entry = RoutingTableEntry(
+            route=routing_entries[next(iter(self.entries))].route,
+            key=self.key, mask=self.mask
+        )
+
+        routing_entries.insert(self.insertion_index, new_entry)
+        aliases[(self.key, self.mask)] = our_aliases = set([])
+
+        # Remove all the old entries
+        offset = 0
+        for i in sorted(self.entries):
+            entry = routing_entries.pop(i - offset)
+            offset += 1
+            km = (entry.key, entry.mask)
+
+            # Add to the alias list
+            our_aliases.update(aliases.pop(km, set([km])))
+
+        return routing_entries, aliases
+
+
+def refine_merge(merge, aliases):
+    """Remove entries from a merge to generate new merges which meet the rules
+    required to allow them to be inserted into the routing table.
+
+    After each modification to the merge a new merge is yielded so that the
+    refining process can be stopped should it not be worth continuing with.
+    The last yielded merge is guaranteed to be valid (this may mean that it is
+    empty).
+
+    Parameters
+    ----------
+    merge : :py:class:`~.Merge`
+        Initial merge to refine.
+    aliases : {(key, mask): {(key, mask), ...}, ...}
+        Map of key-mask pairs to the sets of key-mask pairs that they actually
+        represent.
+
+    Yield
+    -----
+    :py:class:`~.Merge`
+        New set of entries which may be merged.
```

Comment:
It appears you rely on the "goodness" of subsequently produced merges being monotonically decreasing. If so, this should be documented.